book/src/dev/rfcs/0002-parallel-verification.md

```diff
@@ -0,0 +1,335 @@
+* We defer data dependencies until just before the block is committed to the
+  state (see the detaled design RFCs).
+
+# Motivation
+[motivation]: #motivation
+
+Zcash (and Bitcoin) are designed to verify each block in sequence, starting
+from the genesis block. But during the initial sync, and when restarting with
+an older state, this process can be quite slow.
+
+By deferring data dependencies, we can partially verify multiple blocks in
+parallel.
+
+By parallelising block and transaction verification, we can use multithreading
+and batch verification for signatures, proofs, scripts, and hashes.
+
+# Definitions
+[definitions]: #definitions
+
+Blockchain:
+* **chain fork:** Zcash is implemented using a tree of blocks. Each block has a
+                  single previous block, and zero to many next blocks. A chain
+                  fork consists of a tip and all its previous blocks, back to
+                  the genesis block.
+* **genesis:** The root of the tree of blocks is called the genesis block. It has
+               no previous block.
+* **tip:** A block which has no next block is called a tip. Each chain fork can
+           be identified using its tip.
+
+Data:
+* **consensus rule:** A protocol rule which all nodes must apply consistently,
+                      so they can converge on the same chain fork.
+* **context-free:** Consensus rules which do not have a data dependency on
+                    previous blocks.
+* **data dependency:** Information contained in the previous block and its
+                       chain fork, which is required to verify the current block.
+* **semantic verification:** Verifying the consensus rules on the data structures
+                             defined by the protocol.
+* **state:** The set of verified blocks. The state may also cache some dependent
+             data, so that we can efficienty verify subsequent blocks.
+* **structural verification:** Parsing raw bytes into the data structures defined
+                               by the protocol.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+In Zebra, we want to verify blocks in parallel. Some fields can be verified
+straight away, because they don't depend on the output of previous blocks.
+But other fields have **data dependencies**, which means that we need previous
+blocks before we can fully validate them.
+
+If we delay checking some of these data dependencies, then we can do more of
+the verification in parallel.
+
+## Example: BlockHeight
+[block-height]: #block-height
+
+Here's how Zebra can verify the different `BlockHeight` consensus rules in
+parallel:
+
+**Parsing:**
+
+1. Parse the Block into a BlockHeader and a list of transactions.
+
+**Verification - No Data Dependencies:**
+
+2. Check that all BlockHeights are within the range of valid heights.^
+3. Check that the block has exactly one BlockHeight.
+4. Check that the BlockHeight is in the first transaction in the Block.
```

Comment:
I think it would be simpler and more accurate to describe this as:

2. Check that the first input of the first transaction in the block is a coinbase input with a valid block height in its data field.

That's the only block height, by definition. The other consensus rules (only one per block etc) are rules about coinbase inputs, not about block heights, so I don't think they fit in this example.

Reply:
Incidentally, I think that not separating (in the code) the rules about block heights from coinbase inputs led to incorrect implementation of the coinbase checks, fixed in #898.

Reply:
Fixed in 626065adf65e7a3df3be8fbcb1f34399dd7d044c.